source/ada/lsp-ada_documents.adb

```diff
@@ -604,6 +605,314 @@ package body LSP.Ada_Documents is
       end;
    end Diff;
 
+   ------------------
+   -- Diff_Symbols --
+   ------------------
+
+   procedure Diff_Symbols
+     (Self     : Document;
+      Span     : LSP.Messages.Span;
+      New_Text : VSS.Strings.Virtual_String;
+      Edit     : out LSP.Messages.TextEdit_Vector)
+   is
+      use LSP.Types;
+      use LSP.Messages;
+      use VSS.Strings;
+      use VSS.Characters;
+
+      Old_Text  : VSS.Strings.Virtual_String;
+      Old_Lines : VSS.String_Vectors.Virtual_String_Vector;
+      Old_Line  : VSS.Strings.Virtual_String;
+      Old_Length, New_Length : Natural;
+
+      First_Marker : VSS.Strings.Markers.Character_Marker;
+      Last_Marker  : VSS.Strings.Markers.Character_Marker;
+
+   begin
+      Self.Span_To_Markers (Span, First_Marker, Last_Marker);
+
+      Old_Text  := Self.Text.Slice (First_Marker, Last_Marker);
+      Old_Lines := Old_Text.Split_Lines
+        (Terminators     => LSP_New_Line_Function_Set,
+         Keep_Terminator => True);
+      if Natural (Span.last.character) = 0 then
+         --  add empty line to represent the last line if Span.last points
+         --  at the start of it, so we don't have any symbols of it and as
+         --  a result don't have the line at all
+         Old_Lines.Append (Empty_Virtual_String);
+      end if;
+      Old_Line := Old_Lines.Element (Old_Lines.Length);
+
+      Old_Length := Integer (Character_Length (Old_Text));
+      New_Length := Integer (Character_Length (New_Text));
+
+      declare
+         type LCS_Array is array
+           (Natural range 0 .. Old_Length,
+            Natural range 0 .. New_Length) of Integer;
+         type LCS_Array_Access is access all LCS_Array;
+
+         procedure Free is
+           new Ada.Unchecked_Deallocation (LCS_Array, LCS_Array_Access);
+
+         LCS    : LCS_Array_Access := new LCS_Array;
+         Match  : Integer;
+         Delete : Integer;
+         Insert : Integer;
+
+         Old_Char : VSS.Strings.Cursors.Iterators.Characters.
+           Character_Iterator := Old_Text.At_First_Character;
+         New_Char : VSS.Strings.Cursors.Iterators.Characters.
+           Character_Iterator := New_Text.At_First_Character;
+
+         Dummy : Boolean;
+
+         Old_Index, New_Index : Integer;
+
+         Changed_Block_Text : VSS.Strings.Virtual_String;
+         Changed_Block_Span : LSP.Messages.Span := ((0, 0), (0, 0));
+         Span_Set           : Boolean := False;
+
+         --  to calculate span
+         Current_Line_Number : Line_Number := Span.last.line;
+         Old_Lines_Number    : Natural := Old_Lines.Length;
+         Cursor              : VSS.Strings.Cursors.Iterators.Characters.
+           Character_Iterator := Old_Line.After_Last_Character;
+
+         procedure Backward;
+         --  Move old line "cursors" backward
+
+         function Get_Position (Insert : Boolean) return Position;
```

Comment:
Which position? For which span? The comment is not clear at all to me

Reply:
Done